Add ColoredManga

KenjieDec opened the pull request:
Closes #3820

_Please let me know if there's any error when trying to read, since the site is Extremely weird and I almost couldn't understand how it works. 
Thank you._

Checklist:
- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

KenjieDec committed:
Add ColoredManga

- Weird site, very hard to fetch

AwkwardPeak7 commented on src/en/coloredmanga/build.gradle:
```diff
@@ -0,0 +1,11 @@
+apply from: "$rootDir/common.gradle"
+dependencies {
+    implementation project(':lib:dataimage')
+}
```
unused?

AwkwardPeak7 commented on src/en/coloredmanga/src/eu/kanade/tachiyomi/extension/en/coloredmanga/ColoredManga.kt:
```diff
@@ -0,0 +1,395 @@
+
+    private fun getMangas(doc: Document): MutableList<Mangas> {
+        val mangasRaw = doc.selectFirst("script:containsData(\\\"cover)")!!.data()
+        val mangasJson = mangasRaw.replace("\\\"", "\"").replace("\\\\\"", "\\\"").replace("self.__next_f.push([1,\"7:[\"\$\",\"div\",null,{\"className\":\"xl:w-[1280px] w-full min-h-[100vh] font-light flex flex-col items-start justify-start\",\"id\":\"themes_body___1nq3\",\"children\":[\"\$\",\"\$Le\",null,{\"params\":\"\$undefined\",\"data\":[", "{\"data\":[").replace("}]}]}]\\n\"])", "}]}")
```
this can break very easily, either use regex or substrings. also maybe they have an api which can be scraped

KenjieDec replied:
They do have api, I think. But I couldn't understand how it works as the source js uses alphabets and numbers as variables/function name.

AwkwardPeak7 replied:
maybe [this](https://github.com/manga-download/haruneko/pull/657) will help

KenjieDec replied:
Thank you, but can I use replaceBefore and then parse the data json for mangas ( all manga )?:
`val mangasJson = mangasRaw
            .replace(Regex("""\\([\\"])"""), "$1")
            .replaceBefore("\"data\":[", "{")
            .removeSuffix("]}]\\n\"])")`

KenjieDec committed:
Remove unused dependecy

KenjieDec committed:
Apply suggestions, Better chapter name

- Apply AwkwardPeak's suggestions
- Remove genre list

AwkwardPeak7 commented on src/en/coloredmanga/src/eu/kanade/tachiyomi/extension/en/coloredmanga/ColoredManga.kt:
```diff
@@ -0,0 +1,408 @@
+        for (manga in parsedMangas.data) {
+            mangas.add(manga)
+        }
+        return mangas
```
```suggestion
        return mangasJson.parseAs<MangasList>().map { it.data }
```

AwkwardPeak7 commented on src/en/coloredmanga/src/eu/kanade/tachiyomi/extension/en/coloredmanga/ColoredManga.kt:
```diff
@@ -0,0 +1,408 @@
+    // Details
+
+    override fun mangaDetailsParse(response: Response): SManga {
+        val manga = getManga(response)
```
override mangaDetailsRequest to send the multipart request

AwkwardPeak7 commented on src/en/coloredmanga/src/eu/kanade/tachiyomi/extension/en/coloredmanga/ColoredManga.kt:
```diff
@@ -0,0 +1,408 @@
+    // Chapters
+
+    override fun chapterListParse(response: Response): List<SChapter> {
+        val manga = getManga(response)
```
here too

AwkwardPeak7 commented on src/en/coloredmanga/src/eu/kanade/tachiyomi/extension/en/coloredmanga/ColoredManga.kt:
```diff
@@ -0,0 +1,408 @@
+
+    override fun fetchPageList(chapter: SChapter): Observable<List<Page>> {
+        val response = client.newCall(GET(baseUrl + chapter.url.substringBeforeLast("/"), headers)).execute()
+        val manga = getManga(response)
```
here too?

KenjieDec committed:
Appy suggestions, thumbnail fix, chapter name fix

- Apply AwkwardPeak's suggestions
- Thumbnail url might change ( Example: Kuroko's Basketball )
- Chapter name removing all zeros before number fix ( Example: Chapter 001 -> Chapter 1, but Chapter 106 -> Chapter 16 )

AwkwardPeak7 commented on src/en/coloredmanga/src/eu/kanade/tachiyomi/extension/en/coloredmanga/ColoredManga.kt:
```diff
@@ -0,0 +1,417 @@
+
+    // Search
+
+    override fun fetchSearchManga(
+        page: Int,
+        query: String,
+        filters: FilterList,
+    ): Observable<MangasPage> {
+        val response = client.newCall(GET(searchUrl, headers)).execute()
+
+        return Observable.just(
+            searchMangas(response, query, filters),
+        )
+    }
+    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request = popularMangaRequest(page)
+
+    override fun searchMangaParse(response: Response): MangasPage = popularMangaParse(response)
```
you don't need to override fetchSearchManga here

KenjieDec replied:
Hello, is there a way to get the query and filters in `searchMangaParse`? I used `fetchSearchManga` because it has the query and filters to pass to the `searchMangas` function.

AwkwardPeak7 replied:
it's fine then to leave as is

AwkwardPeak7 commented on src/en/coloredmanga/src/eu/kanade/tachiyomi/extension/en/coloredmanga/ColoredManga.kt:
```diff
@@ -0,0 +1,417 @@
+        return getManga(chapter.url.substringBeforeLast("/"))
+    }
+
+    override fun fetchPageList(chapter: SChapter): Observable<List<Page>> {
```
similarly here, you don't need to override fetchPageList

KenjieDec replied:
Hello, is there a way to get the `chapter.url` in `pageListParse`? I need the chapter.url specifically because it contains a segment with the chapter name, which will be removed automatically in `pageListRequest`.

AwkwardPeak7 replied:
you can pass it in url fragment... but it's fine, I didn't notice chapter.url usage previosuly

AwkwardPeak7 commented on src/en/coloredmanga/build.gradle:
```diff
@@ -0,0 +1,8 @@
+ext {
+    extName = 'ColoredManga'
+    extClass = '.ColoredManga'
+    extVersionCode = 1
```
```suggestion
    extVersionCode = 3
```

AwkwardPeak7 commented on src/en/coloredmanga/src/eu/kanade/tachiyomi/extension/en/coloredmanga/ColoredManga.kt:
bump versionId too, since there was an old colored manga extension with same package

AwkwardPeak7 replied:
you didn't push this?

KenjieDec replied:
Sorry for the late commit push, I was waiting for reply on the other suggestions 🙏

KenjieDec committed:
Update build.gradle

Co-authored-by: AwkwardPeak7 <48650614+AwkwardPeak7@users.noreply.github.com>

KenjieDec committed:
Increase versionId
